### PR TITLE
Transform counts fix and log modulo skewness

### PR DESCRIPTION
**Pull request by TuomasBorman**
Hi,

1. now I have disabled some transformation. Following transformations are included:

`transformSamples`:

- log10
- pa
- rank
- relabundance
- hellinger
- clr

`transformFeatures`:

- log10
- pa
- Z

2. I'm not completely sure, how this `transformSamples`/`transformCounts` thing should be constructed. 

My idea was to change `transformCounts` to `transformSamples` and `transformSamples` to `transformCounts`. Then there should be usage section for `transformSamples` not for `transformCounts`. However, when I do this and run `devtools::document()`, I get an error.

```
Error in add_classes_to_exports(ns = nsenv, package = package, exports = exports,  : 
  in ‘mia’ methods for export not found: transformCounts
```

I'm wondering if roxygen is updating namespace correctly, because when I remove e.g. `ZTransform `completely from transformCounts.R, I get similar error. Although, it's highly probable that I'm doing something wrong

3. Should `threshold` parameter be described as additional argument, because it is only used in `pa` transformation? (SImilarly to `estimateDiversity`'s additional arguments.) 

Issue: https://github.com/microbiome/mia/issues/93

### CODE REVIEW

**Comment by TuomasBorman** [comment]
While writing this, I realized that I accidentally created the branch based on my log_modulo_skewness branch. So this PR includes other thing also, sorry for the mess. 

5. Here I implemented, log modulo skewness. Basically, I just copied Leo's code from `microbiome` package. 

6. I addition to log modulo skewness, I put diversity, dominance, and richness indices into alphabetical order, because I thought it is tidyer this way

Issue: https://github.com/microbiome/mia/issues/59

**Comment by FelixErnst** [comment]
Some things are broken in the examples. See the output on the linux test. Cany ou fix this?

File: R/estimateDiversity.R [hunk 1]
```diff
@@ -411,6 +440,39 @@ setMethod("estimateFaith", signature = c(x="TreeSummarizedExperiment", tree="mis
...
+    quantile_point <- quantile(max(mat), quantile)
+    
+    # Tabulate the arithmetic abundance classes. Use the same classes
+    # for all samples for consistency    
+    cutpoints <- c(seq(0, quantile_point, length=num_of_classes), Inf)
+    
+    # Check skewness of the abundance classes for each sample
+    r <- apply(mat, 2, function(x) {
+        .calc_skewness(table(cut(x, cutpoints)))
+    })
```
Comment (FelixErnst) [inline R/estimateDiversity.R]: Is the result of `table` always of the same length? If yes, I think you could construct a matrix and vectorize `.calc_skewness`

File: R/estimateRichness.R [hunk 2]
```diff
@@ -181,7 +183,7 @@ NULL
 #' @export
 setGeneric("estimateRichness",signature = c("x"),
         function(x, abund_values = "counts",
-                  index = c("observed", "chao1", "ace", "hill"),
+                  index = c("ace", "chao1", "hill", "observed"),
```
Comment (FelixErnst) [inline R/estimateRichness.R:186]: These are just rearrangements. What is their purpose?

File: R/transformCounts.R [hunk 3]
```diff
@@ -207,6 +206,16 @@ setMethod("transformCounts", signature = c(x = "SummarizedExperiment"),
                  "different from `abund_values`.",
                  call. = FALSE)
         }
+        # Check method
+        # If method is not single string, user has not specified transform method,
+        # or has given e.g. a vector
+        if(!.is_non_empty_string(method)){
+          stop("'method' must be a non-empty single character value. \n",
+               "Give one method from the following list: \n",
+               "'clr', 'hellinger', 'log10', 'pa', 'rank', 'relabundance'",
```
Comment (FelixErnst) [inline R/transformCounts.R]: This should be removed, because that feature is not tested for in this `if`. I know, this might be helpful for a user at first glance, but essentially it is a crutch.

The user should always to a single point of truth and that is the man page, not an error message.

File: R/transformCounts.R [hunk 4]
```diff
@@ -254,7 +263,17 @@ setMethod("transformFeatures", signature = c(x = "SummarizedExperiment"),
...
                  call. = FALSE)
-            }
+        }
+        # Check method
+        # If method is not single string, user has not specified transform method,
+        # or has given e.g. a vector
+        if(!.is_non_empty_string(method)){
+          stop("'method' must be a non-empty single character value. \n",
+               "Give one method from the following list: \n",
+               "log10', 'pa', 'z'",
```
Comment (FelixErnst) [inline R/transformCounts.R]: See above. It is essentially a crutch and therefore should be removed. The message "Have a look at the man page." would be better if you want to provided resource information

File: tests/testthat/test-5transformCounts.R [hunk 5]
```diff
@@ -166,25 +157,16 @@ test_that("transformCounts", {
...
+        expect_error(mia::transformFeatures(tse, method = "clr"))
 
         ############################# NAMES ####################################
         # Tests that samples have correct names
         expect_equal(colnames(assays(transformCounts(tse, method = "clr", pseudocount = 1))$clr),
                      colnames(assays(tse)$counts))
 
         # Tests that otus have correct names
-        expect_equal(rownames(assays(transformFeatures(tse, method = "hellinger", pseudocount = 1000))$hellinger),
+        expect_equal(rownames(assays(transformFeatures(tse, method = "log10", pseudocount = 1000))$log10),
```
Comment (FelixErnst) [inline tests/testthat/test-5transformCounts.R:169]: Why not test both `hellinger` and `log10`?

File: tests/testthat/test-5transformCounts.R [hunk 6]
```diff
@@ -216,18 +198,8 @@ test_that("transformCounts", {
...
-            # Gets rows from 'rank' table
-            ranks <- assay(tse_rank, "rank")[i,]
-            # Gets rows from 'counts' table, and calculates ranks
-            counts_compare <- assay(tse_rank, "counts")[i,]
-            ranks_compare <- rank(counts_compare, na.last = "keep", ties.method = "first")
-            # Expect that they are equal
-            expect_equal(ranks, ranks_compare)
-        }
+        # Tests transformFeatures, tries to calculate rank. Should be an error.
+        expect_error(mia::transformFeatures(tse, method = "rank"))
```
Comment (FelixErnst) [inline tests/testthat/test-5transformCounts.R:203]: Why was this removed? I think it is an important test.

**Review by FelixErnst — CHANGES_REQUESTED** [review]
Hi Tuomas,

Thanks for the PR.

see the comments. Since this has more than one scope, maybe you want to reflect that in the title of the PR.

Cheers,
Felix

File: R/estimateRichness.R (hunk 2 above)
Comment (TuomasBorman) [reply R/estimateRichness.R:186]: Yes, they are just reordered in alphabetical order. The purpose is that now they are in some kind of order. For me it looks nicer. Also, as a user I would prefer that small things like these are considered

But I can revert that change

File: R/transformCounts.R (hunk 3 above)
Comment (TuomasBorman) [reply R/transformCounts.R]: That's true, methods are not checked in this point. It is just checked that method is not empty

Changed

File: R/transformCounts.R (hunk 4 above)
Comment (TuomasBorman) [reply R/transformCounts.R]: Changed

File: tests/testthat/test-5transformCounts.R (hunk 5 above)
Comment (TuomasBorman) [reply tests/testthat/test-5transformCounts.R:169]: `transformFeatures `does not provide `hellinger` anymore. It provides just `log10`,` pa,` and `z`

File: tests/testthat/test-5transformCounts.R (hunk 6 above)
Comment (TuomasBorman) [reply tests/testthat/test-5transformCounts.R:203]: `transformFeatures` does not provide `rank `anymore. But `rank `transformation for samples (`transformSamples`) are still tested

File: R/estimateDiversity.R (hunk 1 above)
Comment (TuomasBorman) [reply R/estimateDiversity.R]: Vectorized

**Comment by TuomasBorman** [comment]
I modified man page of `transformCounts.` Now it also contains `transformSamples`. However, I'm still not sure about if this is the correct way to do this

File: R/estimateDiversity.R [hunk 7]
```diff
@@ -453,24 +453,34 @@ setMethod("estimateFaith", signature = c(x="TreeSummarizedExperiment", tree="mis
...
-    
-    # Check skewness of the abundance classes for each sample
-    r <- apply(mat, 2, function(x) {
-        .calc_skewness(table(cut(x, cutpoints)))
-    })
-    
+    # Calculates sample-wise frequencies. How many taxa in each interval?
+    freq_table <- table(c(cut(mat, cutpoints)), col(mat))
+    # Calculates the skewness of frequency table. Returns skewness for each sample
+    r <- .calc_skewness(freq_table)
```
Comment (FelixErnst) [inline R/estimateDiversity.R]: I think here is a wrong result returned.

```r
> freq_table <- table(c(cut(mat, cutpoints)), col(mat))
> mia:::.calc_skewness(freq_table)
[1] 2.733111 2.140527 2.691113
```

```r
> mia:::.calc_skewness(apply(mat, 2, function(x) {
+     table(cut(x, cutpoints))
+ }))
[1] 6.490312 5.216445 6.490942
```

I think you need to include a test for correct values.

File: R/estimateDiversity.R [hunk 8]
```diff
@@ -453,24 +453,34 @@ setMethod("estimateFaith", signature = c(x="TreeSummarizedExperiment", tree="mis
...
+    numerator <- numerator/rowSums2(!is.na(x))
+    # Each value is substracted by sample-wise mean, which is raises to the power of 2.
+    # Then the sample-wise sum is taken from these values. 
+    denominator <- rowSums2((x - rowMeans2(x))^2)
+    # Sample-wise sum is divided by number of taxa that are not NA. Then these values
+    # are raised to the power of 3/2.
+    denominator <- (denominator/rowSums2(!is.na(x)))^(3/2)
+    # Result
+    result <- numerator/denominator
+    return(result)
```
Comment (FelixErnst) [inline R/estimateDiversity.R:484]: Nice!

Comment (FelixErnst) [reply R/estimateDiversity.R:484]: Can you include the appropriate import tags as well?

`rowMeans2` and `rowSums2` are both from `matrixStats`

**Review by FelixErnst — CHANGES_REQUESTED** [review]
Looks good, except for the issue with the wrong values.

I changed one chunk to have the man page show `tranfromSamples` before `transfromCounts`. Pleas pull before fixing the issue.

File: R/estimateDiversity.R (hunk 7 above)
Comment (TuomasBorman) [reply R/estimateDiversity.R]: I added tests. However, I got the same result from your example. I also tested `.calc_skewness` with `microbiome'`s value's and they were the same

Comment (FelixErnst) [reply R/estimateDiversity.R]: Hmm, I get the following result

```r
> table(c(cut(mat, cutpoints)), col(mat))
    
      1  2  3
  1  19 16 28
  2   3  6  5
  3   1  3  1
  4   1  2  2
  6   1  2  0
  8   0  1  1
  15  1  0  0
  17  0  3  0
  20  1  0  0
  21  1  0  0
  49  0  0  1
```

but the alternative is of course different

```r
> apply(mat, 2, function(x) {
+     table(cut(x, cutpoints))
+ })
             B  C  D
(0,2.53]    19 16 28
(2.53,5.06]  3  6  5
(5.06,7.59]  1  3  1
(7.59,10.1]  1  2  2
(10.1,12.7]  0  0  0
(12.7,15.2]  1  2  0
(15.2,17.7]  0  0  0
(17.7,20.2]  0  1  1
(20.2,22.8]  0  0  0
(22.8,25.3]  0  0  0
(25.3,27.8]  0  0  0
(27.8,30.4]  0  0  0
(30.4,32.9]  0  0  0
(32.9,35.4]  0  0  0
(35.4,38]    1  0  0
(38,40.5]    0  0  0
(40.5,43]    0  3  0
(43,45.6]    0  0  0
(45.6,48.1]  0  0  0
(48.1,50.6]  1  0  0
(50.6,53.1]  1  0  0
(53.1,55.7]  0  0  0
(55.7,58.2]  0  0  0
(58.2,60.7]  0  0  0
(60.7,63.3]  0  0  0
(63.3,65.8]  0  0  0
(65.8,68.3]  0  0  0
(68.3,70.9]  0  0  0
(70.9,73.4]  0  0  0
(73.4,75.9]  0  0  0
(75.9,78.4]  0  0  0
(78.4,81]    0  0  0
(81,83.5]    0  0  0
(83.5,86]    0  0  0
(86,88.6]    0  0  0
(88.6,91.1]  0  0  0
(91.1,93.6]  0  0  0
(93.6,96.2]  0  0  0
(96.2,98.7]  0  0  0
(98.7,101]   0  0  0
(101,104]    0  0  0
(104,106]    0  0  0
(106,109]    0  0  0
(109,111]    0  0  0
(111,114]    0  0  0
(114,116]    0  0  0
(116,119]    0  0  0
(119,121]    0  0  0
(121,124]    0  0  1
(124,Inf]    0  0  0
```

So this cannot be just happing in my session. `rowMeans` is definitly different with all the zeros in there. Can you check again?

The tests also don't pass for me locally, but they do on GitHub. This weird and probably dangerous

Comment (FelixErnst) [reply R/estimateDiversity.R]: I am still on R 4.0. Is that it?

Comment (TuomasBorman) [reply R/estimateDiversity.R]:
```
> R.version
               _                                        
platform       x86_64-pc-linux-gnu                      
arch           x86_64                                   
os             linux-gnu                                
system         x86_64, linux-gnu                        
status         alpha                                    
major          4                                        
minor          1.0                                      
year           2021                                     
month          04                                       
day            26                                       
svn rev        80229                                    
language       R                                        
version.string R version 4.1.0 alpha (2021-04-26 r80229)
nickname                                                
> table(c(cut(mat, cutpoints)), col(mat))
             
               1  2  3
  (0,2.53]    19 16 28
  (2.53,5.06]  3  6  5
  (5.06,7.59]  1  3  1
  (7.59,10.1]  1  2  2
  (10.1,12.7]  0  0  0
  (12.7,15.2]  1  2  0
  (15.2,17.7]  0  0  0
  (17.7,20.2]  0  1  1
  (20.2,22.8]  0  0  0
  (22.8,25.3]  0  0  0
  (25.3,27.8]  0  0  0
  (27.8,30.4]  0  0  0
  (30.4,32.9]  0  0  0
  (32.9,35.4]  0  0  0
  (35.4,38]    1  0  0
  (38,40.5]    0  0  0
  (40.5,43]    0  3  0
  (43,45.6]    0  0  0
  (45.6,48.1]  0  0  0
  (48.1,50.6]  1  0  0
  (50.6,53.1]  1  0  0
  (53.1,55.7]  0  0  0
  (55.7,58.2]  0  0  0
  (58.2,60.7]  0  0  0
  (60.7,63.3]  0  0  0
  (63.3,65.8]  0  0  0
  (65.8,68.3]  0  0  0
  (68.3,70.9]  0  0  0
  (70.9,73.4]  0  0  0
  (73.4,75.9]  0  0  0
  (75.9,78.4]  0  0  0
  (78.4,81]    0  0  0
  (81,83.5]    0  0  0
  (83.5,86]    0  0  0
  (86,88.6]    0  0  0
  (88.6,91.1]  0  0  0
  (91.1,93.6]  0  0  0
  (93.6,96.2]  0  0  0
  (96.2,98.7]  0  0  0
  (98.7,101]   0  0  0
  (101,104]    0  0  0
  (104,106]    0  0  0
  (106,109]    0  0  0
  (109,111]    0  0  0
  (111,114]    0  0  0
  (114,116]    0  0  0
  (116,119]    0  0  0
  (119,121]    0  0  0
  (121,124]    0  0  1
  (124,Inf]    0  0  0
  > apply(mat, 2, function(x) {
+   table(cut(x, cutpoints))
+ })
             B  C  D
(0,2.53]    19 16 28
(2.53,5.06]  3  6  5
(5.06,7.59]  1  3  1
(7.59,10.1]  1  2  2
(10.1,12.7]  0  0  0
(12.7,15.2]  1  2  0
(15.2,17.7]  0  0  0
(17.7,20.2]  0  1  1
(20.2,22.8]  0  0  0
(22.8,25.3]  0  0  0
(25.3,27.8]  0  0  0
(27.8,30.4]  0  0  0
(30.4,32.9]  0  0  0
(32.9,35.4]  0  0  0
(35.4,38]    1  0  0
(38,40.5]    0  0  0
(40.5,43]    0  3  0
(43,45.6]    0  0  0
(45.6,48.1]  0  0  0
(48.1,50.6]  1  0  0
(50.6,53.1]  1  0  0
(53.1,55.7]  0  0  0
(55.7,58.2]  0  0  0
(58.2,60.7]  0  0  0
(60.7,63.3]  0  0  0
(63.3,65.8]  0  0  0
(65.8,68.3]  0  0  0
(68.3,70.9]  0  0  0
(70.9,73.4]  0  0  0
(73.4,75.9]  0  0  0
(75.9,78.4]  0  0  0
(78.4,81]    0  0  0
(81,83.5]    0  0  0
(83.5,86]    0  0  0
(86,88.6]    0  0  0
(88.6,91.1]  0  0  0
(91.1,93.6]  0  0  0
(93.6,96.2]  0  0  0
(96.2,98.7]  0  0  0
(98.7,101]   0  0  0
(101,104]    0  0  0
(104,106]    0  0  0
(106,109]    0  0  0
(109,111]    0  0  0
(111,114]    0  0  0
(114,116]    0  0  0
(116,119]    0  0  0
(119,121]    0  0  0
(121,124]    0  0  1
(124,Inf]    0  0  0
```

Comment (TuomasBorman) [reply R/estimateDiversity.R]:
```
> R.version
               _                           
platform       x86_64-w64-mingw32          
arch           x86_64                      
os             mingw32                     
system         x86_64, mingw32             
status                                     
major          4                           
minor          0.2                         
year           2020                        
month          06                          
day            22                          
svn rev        78730                       
language       R                           
version.string R version 4.0.2 (2020-06-22)
nickname       Taking Off Again            
> table(c(cut(mat, cutpoints)), col(mat))
    
      1  2  3
  1  19 16 28
  2   3  6  5
  3   1  3  1
  4   1  2  2
  6   1  2  0
  8   0  1  1
  15  1  0  0
  17  0  3  0
  20  1  0  0
  21  1  0  0
  49  0  0  1
> apply(mat, 2, function(x) {
+       table(cut(x, cutpoints))
+    })
             B  C  D
(0,2.53]    19 16 28
(2.53,5.06]  3  6  5
(5.06,7.59]  1  3  1
(7.59,10.1]  1  2  2
(10.1,12.7]  0  0  0
(12.7,15.2]  1  2  0
(15.2,17.7]  0  0  0
(17.7,20.2]  0  1  1
(20.2,22.8]  0  0  0
(22.8,25.3]  0  0  0
(25.3,27.8]  0  0  0
(27.8,30.4]  0  0  0
(30.4,32.9]  0  0  0
(32.9,35.4]  0  0  0
(35.4,38]    1  0  0
(38,40.5]    0  0  0
(40.5,43]    0  3  0
(43,45.6]    0  0  0
(45.6,48.1]  0  0  0
(48.1,50.6]  1  0  0
(50.6,53.1]  1  0  0
(53.1,55.7]  0  0  0
(55.7,58.2]  0  0  0
(58.2,60.7]  0  0  0
(60.7,63.3]  0  0  0
(63.3,65.8]  0  0  0
(65.8,68.3]  0  0  0
(68.3,70.9]  0  0  0
(70.9,73.4]  0  0  0
(73.4,75.9]  0  0  0
(75.9,78.4]  0  0  0
(78.4,81]    0  0  0
(81,83.5]    0  0  0
(83.5,86]    0  0  0
(86,88.6]    0  0  0
(88.6,91.1]  0  0  0
(91.1,93.6]  0  0  0
(93.6,96.2]  0  0  0
(96.2,98.7]  0  0  0
(98.7,101]   0  0  0
(101,104]    0  0  0
(104,106]    0  0  0
(106,109]    0  0  0
(109,111]    0  0  0
(111,114]    0  0  0
(114,116]    0  0  0
(116,119]    0  0  0
(119,121]    0  0  0
(121,124]    0  0  1
(124,Inf]    0  0  0
```

Comment (TuomasBorman) [reply R/estimateDiversity.R]: This problem is related to 4.0

Comment (FelixErnst) [reply R/estimateDiversity.R]: Not exaclty. It is actually the behavior of `c` that changed. From the release notes

```
Using c() to combine a factor with other factors now gives a factor, an ordered factor when combining ordered factors with identical levels.
```

I fixed this in 70ddd9a5
